Bug fixed

AdmiralGorshkov opened the pull request:
Fixed bugs in examples/example1_basic_functionality.jl, and in src/functions.jl. Already tested and now it can load necessary data from given file path, doing calculations as needed, and output graphs for required values.

Commit:
Bug fixed

Fixed bugs in examples/example1_basic_functionality.jl, and in src/functions.jl

ghost commented:
In examples/example1_basic_functionality.jl: running previous code will make an error: MethodError: no method matching pkgdir(::Module, ::String). I modified the way to read the file, so the user can now use the airfoil data file stored anywhere by manually define the path;

In src/functions.jl: running previous code will make these errors: 
1. LoadError: LoadError: syntax: invalid assignment location "; r, n_blades, n_edges" around C:\Users\Administrator\.julia\packages\BEMT\Ac03a\src\functions.jl:79. The bug was fixed by modifying this line of code to assign values separately, instead of deconstructing them all at once;
2. UndefVarError: radius not defined
Stacktrace:
 [1] calculate_vi(rotor::BGeometry{Int64, Float64, Vector{Float64}}, vc::Int64, rpm::Int64, θ::Function, chord::Function, cl::Dierckx.Spline1D, cd::Dierckx.Spline1D)
   @ BEMT C:\Users\Administrator\.julia\packages\BEMT\Ac03a\src\functions.jl:124
By adding radius into function: calculate_vi, this bug was fixed.

mberto79 commented on examples/example1_basic_functionality.jl:
```diff
@@ -2,16 +2,14 @@ using Plots
-
-examples_dir = pkgdir(BEMT, "examples")
-data_dir = joinpath(examples_dir, "data")
+data_dir = "D:\\Document\\BEMT-main\\examples\\data\\example1"
```
@AdmiralGorshkov this change won't work. It is a hard path to your local machine and it will not generalise for other users. 

the approach here may break, so interested to hear about if you have ideas for a more robust approach (without hard paths :smiley:

mberto79 commented on examples/example1_basic_functionality.jl:
```diff
@@ -2,16 +2,14 @@ using Plots
 
 # Define path for xfoil data file and loading
-file_xfoil = joinpath(data_dir, "example1", "NACA001264_xfoil.csv")
+file_xfoil = joinpath(data_dir, "NACA001264_xfoil.csv")
```
@AdmiralGorshkov this follows from the change above so not general enough for other users

mberto79 commented on examples/example1_basic_functionality.jl:
```diff
@@ -2,16 +2,14 @@ using Plots
 
 # Define path for fxlr5 data file and loading
-file_xflr = joinpath(data_dir, "example1", "NACA0012_xflr5.txt")
+file_xflr = joinpath(data_dir, "NACA0012_xflr5.txt")
```
@AdmiralGorshkov this follows from the change above so not general enough for other users

mberto79 commented on src/functions.jl:
```diff
@@ -76,7 +76,7 @@ sigma(chord, radius, nb) = begin
 
 element_performance(rotor, vi, vc, rpm, ρ, cl, cd, θ, chord) = begin
-    (; r, n_blades, n_edges) = rotor
+    r, n_blades, n_edges = rotor.r, rotor.n_blades, rotor.n_edges
```
@AdmiralGorshkov. This deconstruction of the struct should not cause any issues. What version of Julia are you running? This was acceptable syntax since v1.7 (or around there)

mberto79 commented on src/functions.jl:
```diff
@@ -118,8 +118,8 @@ trust_balance(vi, vc, Ω, nb, r, θ, chord, cl, cd) = begin
 
-calculate_vi(rotor, vc, rpm, θ, chord, cl, cd) = begin
-    (; r, radius, n_blades) = rotor
+calculate_vi(radius, rotor, vc, rpm, θ, chord, cl, cd) = begin
```
@AdmiralGorshkov the radius is stored in the rotor structure thus there is no need to pass as a separate arguement

mberto79 reviewed (commented):
@AdmiralGorshkov I don't see how this PR would improve usability. Please see individual comments. It seems like a local (version?) environment issue. 

The paths for the example may need tweaking to work across any device but it tested well on my local machine. The changes to to the de-structuring of the struct (BGeometry) should not have caused any issues. Are you running a recent version of Julia?

Commit:
Incoming file via relative path

Assuming a user downloads the BMt-main directory directly from github and sets it as a working directory in vscode, they just need to put the file in the same place as the sample data file and change the file name in the existing code.

Relative paths are used and the code for the import part of the file is wrapped in function form.

Commit:
Back to original code

It seems like I installed Julia V1.6.9 before, thus using original version of functions.jl is enough.

Commit:
Update functions.jl

AdmiralGorshkov reviewed (commented):
Assuming a user downloads the BEMT main directory directly from github (after they installed the package, in a different position) and sets it as a working directory in vscode, they just need to put the file in the same place as the sample data file and change the file name in the existing code.

Relative paths are used and the code for the import part of the file is wrapped in function form.

ghost commented:
Also I'd like to report a bug: when we set the rpm value higher (more than 2500), the code reports an issue: Warning: Convergence criterion not met!
ERROR: MethodError: Cannot `convert` an object of type Nothing to an object of type Float64

Closest candidates are:
  convert(::Type{T}, ::Gray24) where T<:Real
   @ ColorTypes C:\Users\Administrator\.julia\packages\ColorTypes\1dGw6\src\conversions.jl:114
  convert(::Type{T}, ::Gray) where T<:Real
   @ ColorTypes C:\Users\Administrator\.julia\packages\ColorTypes\1dGw6\src\conversions.jl:113
  convert(::Type{T}, ::T) where T<:Number
   @ Base number.jl:6
  ...

Stacktrace:
 [1] setindex!(A::Vector{Float64}, x::Nothing, i1::Int64)
   @ Base .\array.jl:969
 [2] calculate_vi(rotor::BGeometry{Int64, Float64, Vector{Float64}}, vc::Int64, rpm::Int64, θ::Function, chord::Function, cl::Dierckx.Spline1D, cd::Dierckx.Spline1D)
   @ BEMT d:\Document\BEMT-main\src\functions.jl:128
 [3] top-level scope
   @ d:\Document\BEMT-main\examples\example1_basic_functionality.jl:60
